app/src/main/java/com/example/bottomnav/CAN_Data.java

```diff
@@ -0,0 +1,48 @@
+package com.example.bottomnav;
+
+class CAN_Data {
+    int id;
+    int[] data;
+
+    public CAN_Data(int id, int[] data, int len) {
+        this.id = id;
+        this.data = data;
+    }
+
+    public static CAN_Data decode(String raw) {
+        int id = -1;
+        int len;
+
+        int index = -1;
+        if (raw.charAt(0) == 't' || raw.charAt(0) == 'r') {
+            id = Integer.parseInt(raw.substring(1, 4), 16);
+            index = 4;
+        }
+        else if (raw.charAt(0) == 'T' || raw.charAt(0) == 'R') {
+            id = Integer.parseInt(raw.substring(1, 9), 16);
+            index = 9;
+        }
+        else {
+            return null;
```

Comment:
`nulls` can lead debugging nightmares, because there's nothing in the method signature that indicates that it can be null, so more often then not one might just assume that it'll be non-null.

Where you have a code path that can fail, I recommend encoding the possibility of failure in the type system, such as through an [Optional type](https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html). I believe this is is more common in functional languages, but is more supported elsewhere now (including being part of the Java language)